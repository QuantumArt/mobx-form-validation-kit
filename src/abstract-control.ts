import { action, computed, IReactionDisposer, makeObservable, observable, reaction, runInAction, when } from 'mobx';
import { ControlTypes } from './сontrol-types';
import { ValidationEvent } from './validation-event';
import { ValidationEventTypes } from './validation-event-types';
import { combineErrors, noop } from './utilites';
import { IDictionary } from './idictionary';
import { Delegate } from './delegate';

export type UpdateValidValueHandler<TEntity> = (val: TEntity) => void;
export type ValidatorsFunction<TAbstractControl extends AbstractControl> = (control: TAbstractControl) => Promise<ValidationEvent[]>
export type ControlsCollection = IDictionary<AbstractControl>;

export abstract class AbstractControl {
  /**
   * Type
   * / Тип контрола
   */
  public readonly type: ControlTypes;

  protected inProcessing: boolean;
  /**
   * Validation in progress
   * / В процессе анализа
   */
  abstract processing: boolean;

  private isActiveFunc: () => boolean;

  /**
   * Error checking is disabled (control is always valid)
   * / Проверка ошибок отключена (контрол всегда валиден)
   */
  get disabled(): boolean {
    return !this.active;
  }

  /**
   * Error checking enabled
   * / Проверка ошибок включена
   */
  get active(): boolean {
    return this.isActiveFunc();
  }

  /**
   * Valid
   * / Валидные данные
   */
  get valid(): boolean {
    return !this.invalid;
  }

  /**
   * Invalid
   * / Невалидные данные
   */
  abstract invalid: boolean;

  /**
   * The value has not changed
   * / Значение не изменялось
   */
  get pristine(): boolean {
    return !this.dirty;
  }

  /**
   * Value changed
   * / Значение изменялось
   */
  abstract dirty: boolean;

  /**
   * The field was out of focus
   * / Поле не было в фокусе
   */
  get untouched(): boolean {
    return !this.touched;
  }

  /**
   * The field was in focus
   * / Поле было в фокусе
   */
  abstract touched: boolean;

  /**
   * The field is now in focus
   * / Поле сейчас в фокусе
   */
  abstract focused: boolean;

  private _serverErrors: string[] = [];

  /**
   * Additional (server) errors
   * / Дополнительтные (серверные) ошибки
   */
  public get serverErrors(): string[] {
    return this._serverErrors;
  }

  /**
   * Additional (server) errors
   * / Дополнительтные (серверные) ошибки
   */
  public set serverErrors(value: string[]) {
    this._serverErrors = value || [];
  }

  /**
   * Errors list
   * / Список ошибок
   */
  public errors: ValidationEvent[] = [];

  /**
   *  The field contains errors
   * / Присутствуют ошибки
   */
  public hasErrors() {
    return (!!this.errors && this.errors.length > 0) || (!!this._serverErrors && this._serverErrors.length > 0);
  }

  /**
   * Warnings messages list
   * / Список сообщений с типом "Внимание"
   */
  public warnings: ValidationEvent[] = [];

  /**
   *  The field contains warnings messages
   * / Присутствуют сообщения с типом "Внимание"
   */
  public hasWarnings() {
    return !!this.warnings && this.warnings.length > 0;
  }

  /**
   * Informations messages list
   * / Сообщения с типом "Информационные сообщения"
   */
  public informationMessages: ValidationEvent[] = [];

  /**
   *  The field contains informations messages
   * / Присутствуют сообщения с типом "Информационные сообщения"
   */
  public hasInformationMessages() {
    return !!this.informationMessages && this.informationMessages.length > 0;
  }

  /**
   * Successes messages list
   * / Сообщения с типом "успешная валидация"
   */
  public successes: ValidationEvent[] = [];

  /**
   *  The field contains successes
   * / Присутствуют сообщения с типом "успешная валидация"
   */
  public hasSuccesses() {
    return !!this.successes && this.successes.length > 0;
  }

  /**
   * Max message level
   * / Максимальный уровень сообщения
   */
  get maxEventLevel(): ValidationEventTypes {
    if (this.hasErrors()) return ValidationEventTypes.Error;
    if (this.hasWarnings()) return ValidationEventTypes.Warning;
    if (this.hasInformationMessages()) return ValidationEventTypes.Info;
    return ValidationEventTypes.Success;
  }

  /**
   * Set marker "value changed"
   * / Изменяет состояния маркета "данные изменены"
   */
  abstract setDirty(dirty: boolean): this;

  /**
   * Set marker "field was out of focus"
   * / Изменяет состояния маркета "значение было в фокусе"
   */
  abstract setTouched(touched: boolean): this;

  /**
   * Returns a datastructure fully representing controls value
   */
  abstract getValue(): any;

  /**
   * Field for transferring additional information
   * / Поле для передачи дополнительной информации (в логике не участвует) 
   */
  public additionalData: any;

  public element: HTMLElement | null = null;

  /**
   * Callback function of on change
   * / Сообщает факт изменения данных
   */
  public onChange: Delegate<AbstractControl> = new Delegate<AbstractControl>();

  constructor(
    /**
     * Function enable validation by condition (always enabled by default)
     * / Функция включение валидаций по условию (по умолчанию включено всегда)
     */
    activate: (() => boolean) | null = null,
    additionalData: any,
    type: ControlTypes
  ) {
    makeObservable<AbstractControl, 'inProcessing' | '_serverErrors' | 'onValidation'>(this, {
      inProcessing: observable,

      disabled: computed,
      active: computed,
      valid: computed,
      pristine: computed,
      untouched: computed,

      _serverErrors: observable,
      serverErrors: computed,

      errors: observable.ref,
      warnings: observable.ref,
      informationMessages: observable.ref,
      successes: observable.ref,
      maxEventLevel: computed,

      additionalData: observable,

      onValidation: action
    });

    this.inProcessing = false;
    this.isActiveFunc = activate === null ? () => true : activate;
    this.additionalData = additionalData;
    this.type = type;
  }

  /**
   * Dispose (call in unmount react control)
   * / Вызвать при удалении контрола
   */
  dispose(): void {
    this.onChange.dispose();
    for (const reactionOnValidator of this.reactionOnValidatorDisposers) {
      reactionOnValidator();
    }
  }

  /**
   * Get error by key
   * / Получить ошибку по ключу
   */
  public error = (key: string): ValidationEvent | undefined => {
    return this.errors.find(err => err.key === key);
  };

  private newRequestValidation: number = 0;
  private lastValidators: ValidatorsFunction<any>[] = [];
  private lastValidationFunction = noop;
  protected reactionOnValidatorDisposers: IReactionDisposer[] = [];

  protected onValidation = async <TAbstractControl extends AbstractControl>(
    validators: ValidatorsFunction<TAbstractControl>[],
    onValidationFunction: () => void,
    afterCheck: () => void,
  ): Promise<void> => {
    const haveRequestValidation: boolean = this.newRequestValidation !== 0;
    this.newRequestValidation++;
    this.lastValidators = validators;
    this.lastValidationFunction = onValidationFunction;
    if (haveRequestValidation) {
      return;
    }
    let groupErrors: ValidationEvent[][];
    let oldRequestValidation = 0;
    do {
      oldRequestValidation = this.newRequestValidation;
      this.reactionOnValidatorDisposers.forEach(r => r());
      this.reactionOnValidatorDisposers = [];
      if (this.active) {
        const errorsPromises = this.lastValidators.map(validator => {
          let isFirstReaction = true;
          return new Promise<ValidationEvent[]>(resolve =>
            this.reactionOnValidatorDisposers.push(
              reaction(() => {
                let result;
                if (isFirstReaction) {
                  result = validator(this).then(resolve);
                }
                isFirstReaction = false;
                return result;
              }, this.lastValidationFunction),
            ),
          );
        });
        groupErrors = await Promise.all(errorsPromises);
      } else {
        groupErrors = [];
      }
    } while (oldRequestValidation !== this.newRequestValidation);
    this.newRequestValidation = 0;
    const events = groupErrors && groupErrors.length > 0 ? combineErrors(groupErrors) : [];
    runInAction(() => {
      this.errors = events.filter(e => e.type === ValidationEventTypes.Error);
      this.warnings = events.filter(e => e.type === ValidationEventTypes.Warning);
      this.informationMessages = events.filter(e => e.type === ValidationEventTypes.Info);
      this.successes = events.filter(e => e.type === ValidationEventTypes.Success);

      afterCheck();
    });
  };

  /**
   * Waiting for end of validation
   * Ожидание окончания проверки
   */
  public wait(): Promise<void> {
    return when(() => !this.processing);
  }

  public abstract executeAsyncValidation(validator: (control: this) => Promise<ValidationEvent[]>): Promise<ValidationEvent[]>;

  public abstract runInAction(action: () => void): void;

  protected baseExecuteAsyncValidation = (
    validator: (control: this) => Promise<ValidationEvent[]>,
    onValidationFunction: () => void,
  ): Promise<ValidationEvent[]> => {
    let isFirstReaction = true;
    return new Promise<ValidationEvent[]>(resolve =>
      this.reactionOnValidatorDisposers.push(
        reaction(() => {
          let result;
          if (isFirstReaction) {
            result = validator(this).then(resolve);
          }
          isFirstReaction = false;
          return result;
        }, onValidationFunction),
      ),
    );
  };
}
