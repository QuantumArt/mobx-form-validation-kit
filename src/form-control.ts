import { action, computed, IReactionDisposer, makeObservable, observable, reaction } from 'mobx';
import { AbstractControl, ControlTypes, noop, UpdateValidValueHandler, ValidationEvent, ValidatorsFunction } from './internal';

interface OptionsFormControl<TEntity> {
  /**
   * Validations
   * / Валидациии
   */
  validators?: ValidatorsFunction<FormControl<TEntity>>[];
  /**
   * Function enable validation by condition (always enabled by default)
   * / Функция включение валидаций по условию (по умолчанию включено всегда)
   */
  activate?: (() => boolean) | null;
  /**
   * Additional information
   * / Блок с дополнительной информацией
   */
  additionalData?: any;
  /**
   * Callback always when value changes
   * / Срабатывает всегда при изменении значения
   */
  onChangeValue?: UpdateValidValueHandler<TEntity> | null;
  /**
   * Callback get last valid value
   * / Передает последние валидное значение
   */
  onChangeValidValue?: UpdateValidValueHandler<TEntity> | null;
  /**
   * Invoke `onChangeValidValue` when `FormControl` is created.
   * / Вызвать `onChangeValidValue` при создании `FormControl`.
   * @example
   * const model = observable({ value: 123 });
   * new FormControl(
   *   () => model.value,
   *   [],
   *   value => { console.log({ value }); },
   *   { callSetterOnInitialize: true }
   * ); // then we see { value: 123 } in console immediately
   */
  callSetterOnInitialize?: boolean;
  /**
   * Invoke `onChangeValidValue` when value-getter that passed as first argument changes its underlying value.
   * / Вызывать `onChangeValidValue` при каждом изменении результата функции-геттера из первого аргумента.
   * @default false
   * @example
   * const model = observable({ value: 123 });
   * new FormControl(
   *   () => model.value,
   *   [],
   *   value => { console.log({ value }); },
   *   { callSetterOnReinitialize: true }
   * );
   * model.value = 456; // then we see { value: 456 } in console
   */
  callSetterOnReinitialize?: boolean;
}

export class FormControl<TEntity = string> extends AbstractControl {
  private reactionOnValueGetterDisposer: IReactionDisposer | undefined;
  private reactionOnInternalValueDisposer: IReactionDisposer | undefined;
  private readonly reactionOnIsActiveDisposer: IReactionDisposer;
  private readonly reactionOnIsDirtyDisposer: IReactionDisposer;
  private readonly reactionOnIsFocusedDisposer: IReactionDisposer;
  private readonly validators: ValidatorsFunction<FormControl<TEntity>>[];
  private readonly setValidValue: UpdateValidValueHandler<TEntity>;
  private readonly onChangeValue: UpdateValidValueHandler<TEntity>;
  private readonly callSetterOnInitialize: boolean;
  private readonly callSetterOnReinitialize: boolean;

  private isInitializedValue: boolean = false;
  private isInitializedActived: boolean = false;
  private get isInitialized(): boolean {
    return this.isInitializedValue || this.isInitializedActived;
  }

  get processing(): boolean {
    return this.inProcessing;
  }

  // @ts-ignore: internalValue is always initialized in this.setInitialValue() call
  private internalValue: TEntity;

  public get value(): TEntity {
    return this.internalValue;
  }

  public set value(value: TEntity) {
    this.internalValue = value;
  }

  get invalid(): boolean {
    return this.active && (this.errors.length > 0 || this.serverErrors.length > 0);
  }

  private isDirty: boolean = false;

  get dirty(): boolean {
    return this.isDirty;
  }

  private isTouched: boolean = false;

  get touched(): boolean {
    return this.isTouched;
  }

  private isFocused: boolean = false;

  get focused(): boolean {
    return this.isFocused;
  }

  constructor(
    /**
     * Initializing valueI
     * / Инициализирующие значение или его getter
     */
    valueOrGetter: TEntity | (() => TEntity),
    /**
     * Options
     * / Опции
     */
    options: OptionsFormControl<TEntity> = {},
  ) {
    super(options.activate ?? null, options.additionalData, ControlTypes.Control);
    makeObservable<FormControl<TEntity>, 'internalValue' | 'isDirty' | 'isTouched' | 'isFocused' | 'checkInternalValue'>(this, {
      processing: computed,

      internalValue: observable,
      value: computed,

      invalid: computed,

      isDirty: observable,
      dirty: computed,

      isTouched: observable,
      touched: computed,

      isFocused: observable,
      focused: computed,

      setDirty: action,
      setTouched: action,
      setFocused: action,
      checkInternalValue: action,
    });

    this.validators = options.validators ?? [];
    this.setValidValue = options.onChangeValidValue ?? noop;
    this.onChangeValue = options.onChangeValue ?? noop;
    this.additionalData = options.additionalData ?? null;
    this.callSetterOnInitialize = options.callSetterOnInitialize ?? true;
    this.callSetterOnReinitialize = options.callSetterOnReinitialize ?? false;

    this.reactionOnIsActiveDisposer = reaction(
      () => this.active,
      (active) => {
        this.serverErrors = [];
        if (active) {
          this.checkInternalValue(this.isInitialized || this.callSetterOnInitialize);
        }
        this.isInitializedActived = true;
        this.onChange.call(this);
      },
    );

    this.reactionOnIsDirtyDisposer = reaction(
      () => this.dirty,
      (isDirty: boolean) => {
        if (isDirty) {
          this.serverErrors = [];
        }
      },
    );

    this.reactionOnIsFocusedDisposer = reaction(
      () => this.focused,
      (isFocused: boolean) => {
        if (!isFocused) {
          this.serverErrors = [];
        }
      },
    );

    this.setInitialValue(valueOrGetter);
  }

  public setInitialValue = (valueOrGetter: TEntity | (() => TEntity)) => {
    const valueGetter = valueOrGetter instanceof Function ? valueOrGetter : () => valueOrGetter;

    this.reactionOnValueGetterDisposer && this.reactionOnValueGetterDisposer();

    this.reactionOnValueGetterDisposer = reaction(
      valueGetter,
      initialValue => {
        this.reactionOnInternalValueDisposer && this.reactionOnInternalValueDisposer();
        this.internalValue = initialValue;
        this.reactionOnInternalValueDisposer = reaction(
          () => this.internalValue,
          () => {
            this.onChangeValue(this.internalValue);
            this.isDirty = true;
            this.serverErrors = [];
            this.onChange.call(this);
            this.checkInternalValue(true);
          },
        );
        this.onChange.call(this);
        this.checkInternalValue(this.isInitialized ? this.callSetterOnReinitialize : this.callSetterOnInitialize);
        this.isInitializedValue = true;
      },
      { fireImmediately: true },
    );

    return this;
  };

  public executeAsyncValidation = (validator: (control: this) => Promise<ValidationEvent[]>): Promise<ValidationEvent[]> =>
    this.baseExecuteAsyncValidation(validator, () => this.checkInternalValue(true));

  /**
  * Set marker "Value has changed" 
  * / Установить маркер "Значение изменилось"
  */
  public setDirty = (dirty: boolean) => {
    this.isDirty = dirty;
    return this;
  };

  /**
   * Set marker "field was in focus" 
   * / Установить маркер "Поле было в фокусе"
   */
  public setTouched = (touched: boolean) => {
    this.isTouched = touched;
    return this;
  };

  public setFocused = (focused: boolean) => {
    this.isFocused = focused;
    return this;
  };

  public dispose = (): void => {
    super.dispose();
    this.reactionOnValueGetterDisposer && this.reactionOnValueGetterDisposer();
    this.reactionOnInternalValueDisposer && this.reactionOnInternalValueDisposer();
    this.reactionOnIsActiveDisposer();
    this.reactionOnIsDirtyDisposer();
    this.reactionOnIsFocusedDisposer();
  };

  public runInAction(action: () => void): void {
    this.reactionOnValidatorDisposers.push(
      reaction(
        () => action(),
        () => this.checkInternalValue(true),
      ),
    );
  };

  private checkInternalValue = (shouldCallSetter: boolean) => {
    this.inProcessing = true;
    this.serverErrors = [];
    this.onValidation(this.validators, () => this.checkInternalValue(true), () => {
      if (shouldCallSetter && this.setValidValue && this.errors.length === 0) {
        this.setValidValue(this.internalValue);
      }
      this.inProcessing = false;
    });
  };
}
