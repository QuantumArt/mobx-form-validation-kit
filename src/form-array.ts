import { action, computed, IReactionDisposer, observable, reaction } from 'mobx';
import { AbstractControl, ValidatorsFunction } from './abstract-control';
import { FormAbstractGroup } from './form-abstract-group';
import { ControlTypes } from './сontrol-types';
import { ValidationEvent } from './validation-event';

export interface IOptionsFormArray<TAbstractControl extends AbstractControl> {
  /**
    * Validations
    * Валидациии
    */
  validators?: ValidatorsFunction<FormArray<TAbstractControl>>[],
  /**
    * Additional information
    * Блок с дополнительной информацией
    */
  additionalData?: any;
  /**
    * Function enable validation by condition (always enabled by default)
    * / Функция включение валидаций по условию (по умолчанию включено всегда)
    */
  activate?: (() => boolean) | null;
}

export class FormArray<TAbstractControl extends AbstractControl> extends FormAbstractGroup {
  private readonly reactionOnIsActiveDisposer: IReactionDisposer;

  @observable
  private controls: TAbstractControl[];

  @computed
  public get length(): number {
    return this.controls.length;
  }

  private readonly validators: ValidatorsFunction<FormArray<TAbstractControl>>[] = [];

  constructor(
    /**
      * Сontrols
      * / Контролы
      */
    controls: TAbstractControl[],
    /**
     * Options
     * / Опции
     */
    options: IOptionsFormArray<TAbstractControl> = {},
  ) {
    super(options.activate ?? null, options.additionalData, ControlTypes.Array);
    this.inProcessing = false;
    this.controls = controls ?? [];
    this.validators = options.validators ?? [];

    this.reactionOnIsActiveDisposer = reaction(
      () => this.active,
      () => {
        this.checkArrayValidations();
        this.onChange.call(this);
      },
    );

    for (const control of this.controls) {
      control.onChange.addListen(() => {
        this.serverErrors = [];
        this.checkArrayValidations();
        this.onChange.call(this);
      });
    }

    this.checkArrayValidations();
  }

  public get(index: number): TAbstractControl {
    return this.controls[index];
  }

  public dispose = (): void => {
    super.dispose();
    this.reactionOnIsActiveDisposer();
    for (const control of this.controls) {
      control.dispose();
    }
  };

  public executeAsyncValidation = (validator: (control: this) => Promise<ValidationEvent[]>): Promise<ValidationEvent[]> =>
    this.baseExecuteAsyncValidation(validator, () => this.checkArrayValidations());

  @action
  private checkArrayValidations = () => {
    this.inProcessing = true;
    this.serverErrors = [];
    this.onValidation(this.validators, this.checkArrayValidations, () => this.inProcessing = false);
  };

  public runInAction(action: () => void): void {
    this.reactionOnValidatorDisposers.push(
      reaction(
        () => action(),
        () => this.checkArrayValidations()
      )
    );
  };

  /**
   * Removes the last element from an array and returns it.
   */
  @action
  public pop = (): TAbstractControl | undefined => {
    const removeControl = this.controls.pop();
    this.onChange.call(this);
    return removeControl;
  };

  /**
   * Appends new elements to an array, and returns the new length of the array.
   * @param items New elements of the Array.
   */
  @action
  public push = (...items: TAbstractControl[]): number => {
    const newIndex = this.controls.push(...items);
    this.onChange.call(this);
    return newIndex;
  };

  /**
   * Combines two or more arrays.
   * @param items Additional items to add to the end of array1.
   */
  @action
  public concat = (...items: (TAbstractControl | ConcatArray<TAbstractControl>)[]): TAbstractControl[] => {
    return this.controls.concat(...items);
  };

  /**
   * Combines two or more arrays.
   * @param items Additional items to add to the end of array1.
   */
  @action
  public clear = () => {
    this.controls = [];
    this.onChange.call(this);
  };

  /**
   * Reverses the elements in an Array.
   */
  @action
  public reverse = (): TAbstractControl[] => {
    return this.controls.reverse();
  };

  /**
   * Removes the first element from an array and returns it.
   */
  @action
  public shift = (): TAbstractControl | undefined => {
    return this.controls.shift();
  };

  /**
   * Returns a section of an array.
   * @param start The beginning of the specified portion of the array.
   * @param end The end of the specified portion of the array.
   */
  public slice = (start?: number, end?: number): TAbstractControl[] => {
    return this.controls.slice(start, end);
  };

  /**
   * Sorts an array.
   * @param compareFn The name of the function used to determine the order of the elements. If omitted, the elements are sorted in ascending, ASCII character order.
   */
  @action
  public sort = (compareFn?: (a: TAbstractControl, b: TAbstractControl) => number) => {
    return this.controls.slice().sort(compareFn);
  };

  /**
   * Removes elements from an array and, if necessary, inserts new elements in their place, returning the deleted elements.
   * @param start The zero-based location in the array from which to start removing elements.
   * @param deleteCount The number of elements to remove.
   */
  public splice = (start: number, deleteCount?: number): TAbstractControl[] => {
    return this.controls.splice(start, deleteCount);
  };

  /**
   * Inserts new elements at the start of an array.
   * @param items  Elements to insert at the start of the Array.
   */
  public unshift = (...items: TAbstractControl[]): number => {
    return this.controls.unshift(...items);
  };

  /**
   * Returns the index of the first occurrence of a value in an array.
   * @param searchElement The value to locate in the array.
   * @param fromIndex The array index at which to begin the search. If fromIndex is omitted, the search starts at index 0.
   */
  public indexOf = (searchElement: TAbstractControl, fromIndex?: number): number => {
    return this.controls.indexOf(searchElement, fromIndex);
  };

  /**
   * Returns the index of the last occurrence of a specified value in an array.
   * @param searchElement The value to locate in the array.
   * @param fromIndex The array index at which to begin the search. If fromIndex is omitted, the search starts at the last index in the array.
   */
  public lastIndexOf = (searchElement: TAbstractControl, fromIndex?: number): number => {
    return this.controls.lastIndexOf(searchElement, fromIndex);
  };

  /**
   * Determines whether all the members of an array satisfy the specified test.
   * @param callbackfn A function that accepts up to three arguments. The every method calls the callbackfn function for each element in array1 until the callbackfn returns false, or until the end of the array.
   * @param thisArg An object to which the this keyword can refer in the callbackfn function. If thisArg is omitted, undefined is used as the this value.
   */
  public every = (callbackfn: (value: TAbstractControl, index: number, array: TAbstractControl[]) => unknown, thisArg?: any): boolean => {
    return this.controls.every(callbackfn, thisArg);
  };

  /**
   * Determines whether the specified callback function returns true for any element of an array.
   * @param callbackfn A function that accepts up to three arguments. The some method calls the callbackfn function for each element in array1 until the callbackfn returns true, or until the end of the array.
   * @param thisArg An object to which the this keyword can refer in the callbackfn function. If thisArg is omitted, undefined is used as the this value.
   */
  public some = (callbackfn: (value: TAbstractControl, index: number, array: TAbstractControl[]) => unknown, thisArg?: any): boolean => {
    return this.controls.some(callbackfn, thisArg);
  };

  /**
   * Performs the specified action for each element in an array.
   * @param callbackfn  A function that accepts up to three arguments. forEach calls the callbackfn function one time for each element in the array.
   * @param thisArg  An object to which the this keyword can refer in the callbackfn function. If thisArg is omitted, undefined is used as the this value.
   */
  public forEach = (callbackfn: (value: TAbstractControl, index: number, array: TAbstractControl[]) => void, thisArg?: any): void => {
    return this.controls.forEach(callbackfn, thisArg);
  };

  /**
   * Calls a defined callback function on each element of an array, and returns an array that contains the results.
   * @param callbackfn A function that accepts up to three arguments. The map method calls the callbackfn function one time for each element in the array.
   * @param thisArg An object to which the this keyword can refer in the callbackfn function. If thisArg is omitted, undefined is used as the this value.
   */
  public map = <U>(callbackfn: (value: TAbstractControl, index: number, array: TAbstractControl[]) => U, thisArg?: any): U[] => {
    return this.controls.map(callbackfn, thisArg);
  };

  /**
   * Returns the elements of an array that meet the condition specified in a callback function.
   * @param callbackfn A function that accepts up to three arguments. The filter method calls the callbackfn function one time for each element in the array.
   * @param thisArg An object to which the this keyword can refer in the callbackfn function. If thisArg is omitted, undefined is used as the this value.
   */
  public filter = (callbackfn: (value: TAbstractControl, index: number, array: TAbstractControl[]) => unknown, thisArg?: any): TAbstractControl[] => {
    return this.controls.filter(callbackfn, thisArg);
  };

  /**
   * Calls the specified callback function for all the elements in an array. The return value of the callback function is the accumulated result, and is provided as an argument in the next call to the callback function.
   * @param callbackfn A function that accepts up to four arguments. The reduce method calls the callbackfn function one time for each element in the array.
   * @param initialValue If initialValue is specified, it is used as the initial value to start the accumulation. The first call to the callbackfn function provides this value as an argument instead of an array value.
   */
  public reduce = <U = TAbstractControl>(
    callbackfn: (previousValue: U, currentValue: TAbstractControl, currentIndex: number, array: TAbstractControl[]) => U,
    initialValue: U,
  ): U => {
    return this.controls.reduce(callbackfn, initialValue);
  };

  /**
   * Calls the specified callback function for all the elements in an array, in descending order. The return value of the callback function is the accumulated result, and is provided as an argument in the next call to the callback function.
   * @param callbackfn A function that accepts up to four arguments. The reduceRight method calls the callbackfn function one time for each element in the array.
   * @param initialValue If initialValue is specified, it is used as the initial value to start the accumulation. The first call to the callbackfn function provides this value as an argument instead of an array value.
   */
  public reduceRight = <U>(
    callbackfn: (previousValue: U, currentValue: TAbstractControl, currentIndex: number, array: TAbstractControl[]) => U,
    initialValue: U,
  ): U => {
    return this.controls.reduceRight(callbackfn, initialValue);
  };

  protected *getControls(): IterableIterator<AbstractControl> {
    for (const control of this.controls) {
      yield control;
    }
  }
}
