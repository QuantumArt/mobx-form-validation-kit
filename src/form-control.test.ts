import { observable } from 'mobx';
import { ControlsCollection } from './abstract-control';
import { FormArray } from './form-array';
import { FormControl } from './form-control';
import { FormGroup } from './form-group';
import { ValidationEvent } from './validation-event';
import { ValidationEventTypes } from './validation-event-types';
import { maxValueValidator, minValueValidator, patternValidator, requiredValidator, wrapperActivateValidation, wrapperSequentialCheck } from './validators';

describe('FormControl', () => {
  it('should not call setter when initialized by default', async () => {
    const setter = jest.fn<void, [string]>();

    const form = new FormGroup({
      field: new FormControl<string>('test', {
        validators: [requiredValidator()],
        onChangeValidValue: setter,
        callSetterOnInitialize: false,
      }),
    });
    await form.wait();

    expect(setter).not.toBeCalled();
  });

  it('should call setter once when value is changed', async () => {
    const setter = jest.fn<void, [string]>();

    const form = new FormGroup({
      field: new FormControl<string>('test', {
        validators: [requiredValidator()],
        onChangeValidValue: setter,
        callSetterOnInitialize: false,
      }),
    });
    await form.wait();

    form.controls.field.value = 'qwerty';
    await form.wait();

    expect(setter).toBeCalledTimes(1);
    expect(setter).toBeCalledWith('qwerty');
  });

  it('should reflect initial value getter changes', async () => {
    const model = observable({ field: 'test' });

    const setter = jest.fn<void, [string]>();

    const form = new FormGroup({
      field: new FormControl<string>(() => model.field, {
        validators: [requiredValidator()],
        onChangeValidValue: setter,
      }),
    });
    await form.wait();

    expect(form.controls.field.value).toEqual('test');

    model.field = 'qwerty';
    await form.wait();

    expect(form.controls.field.value).toEqual('qwerty');
  });

  it('should not call setter when reinitialized by default', async () => {
    const model = observable({ field: 'test' });

    const setter = jest.fn<void, [string]>();

    const form = new FormGroup({
      field: new FormControl<string>(() => model.field, {
        validators: [requiredValidator()],
        onChangeValidValue: setter,
        callSetterOnInitialize: false
      }),
    });
    await form.wait();

    model.field = 'qwerty';
    await form.wait();

    expect(setter).not.toBeCalled();
  });

  it('should not call setter when activated during initialization', async () => {
    const primarySetter = jest.fn<void, [string]>();
    const dependentSetter = jest.fn<void, [string]>();

    interface IForm extends ControlsCollection {
      primaryField: FormControl<string>;
      dependentField: FormControl<string>;
    }

    class Component {
      @observable form: FormGroup<IForm> = new FormGroup({
        primaryField: new FormControl<string>('foo', {
          validators: [requiredValidator()],
          onChangeValidValue: primarySetter,
          callSetterOnInitialize: false,
        }),
        dependentField: new FormControl<string>('bar', {
          validators: [requiredValidator()],
          onChangeValidValue: dependentSetter,
          activate: () => this.form && this.form.controls.primaryField.value === 'foo',
          callSetterOnInitialize: false,
        }),
      });
    }

    const component = new Component();

    await component.form.wait();

    expect(primarySetter).not.toBeCalled();
    expect(dependentSetter).not.toBeCalled();
  });

  it('should call setter once when activated after initialization', async () => {
    const primarySetter = jest.fn<void, [number]>();
    const dependentSetter = jest.fn<void, [string]>();

    interface IForm extends ControlsCollection {
      primaryField: FormControl<number>;
      dependentField: FormControl<string>;
    }

    class Component {
      @observable form: FormGroup<IForm> = new FormGroup({
        primaryField: new FormControl<number>(123, {
          validators: [requiredValidator() as any],
          onChangeValidValue: primarySetter,
          callSetterOnInitialize: false,
        }),
        dependentField: new FormControl<string>('bar', {
          validators: [requiredValidator()],
          onChangeValidValue: dependentSetter,
          activate: () => this.form && this.form.controls.primaryField.value === 456,
          callSetterOnInitialize: false,
        }),
      });
    }

    const component = new Component();
    await component.form.wait();

    component.form.controls.primaryField.value = 456;
    await component.form.wait();

    expect(primarySetter).toBeCalledTimes(1);
    expect(primarySetter).toBeCalledWith(456);
    expect(dependentSetter).toBeCalledTimes(1);
    expect(dependentSetter).toBeCalledWith('bar');
  });

  it('test array', async () => {
    const form = new FormArray(
      [new FormControl<string>(''), new FormControl<string>('')],
      {
        validators:
          [
            async (array: FormArray<FormControl<string>>): Promise<ValidationEvent[]> => {
              if (array.some(i => !!i.value)) {
                return [
                  {
                    message: '',
                    type: ValidationEventTypes.Error,
                  },
                ];
              }
              return [];
            },
          ]
      },
    );
    await form.wait();
    expect(form.valid).toBe(true);

    form.get(1).value = 'test';
    await form.wait();

    expect(form.valid).toBe(false);
  });

  it('wrapper on array', async () => {
    const form = new FormArray(
      [new FormControl<string>(''), new FormControl<string>('')],
      {
        validators:
          [wrapperSequentialCheck([wrapperActivateValidation(() => true, [])])]
      },
    );

    await form.wait();

    expect(form.valid).toBe(true);
  });

  it('minValue and maxValue', async () => {
    const form = new FormGroup({
      count: new FormControl<number>(0, {
        validators: [minValueValidator<number>(1, 'Должна быть оценка'), maxValueValidator<number>(5, 'Должна быть оценка')]
      }),
    });

    await form.wait();
    expect(form.valid).toBe(false);

    form.controls.count.value = 3;

    await form.wait();
    expect(form.valid).toBe(true);
  });

  it('activate validation', async () => {
    class Component {
      @observable activateValidation: boolean = false;
    }

    const component = new Component();
    const form = new FormGroup({
      str: new FormControl<string>('', {
        validators: [wrapperActivateValidation(() => component.activateValidation, [requiredValidator()])]
      }),
    });

    await form.wait();
    expect(form.valid).toBe(true);

    component.activateValidation = true;

    await form.wait();
    expect(form.valid).toBe(false);
  });

  it('wrapper sequential check', async () => {
    const form = new FormGroup({
      date: new FormControl<string | null>('10.10.1010', {
        validators: [
          wrapperSequentialCheck([requiredValidator(), patternValidator(/^\d\d.\d\d.\d\d\d\d$/, 'Введите дату в формате "дд.мм.гггг"')]
          ),
        ]
      }),
    });

    await form.wait();
    expect(form.valid).toBe(true);

    form.controls.date.value = '10101010';

    await form.wait();
    expect(form.valid).toBe(false);
  });

  it('wrapper on array', async () => {
    const form = new FormArray([new FormControl<string>(''), new FormControl<string>('')], {
      validators: [wrapperSequentialCheck([wrapperActivateValidation(() => true, [])])],
    });

    await form.wait();
    expect(form.map(e => e).length).toBe(2);
    expect(form.allControls().length).toBe(2);

    expect(form.valid).toBe(true);
  });
});
