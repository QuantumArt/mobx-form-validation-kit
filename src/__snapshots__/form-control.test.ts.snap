// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormControl getValue allows to get all values recursivelly 1`] = `
Object {
  "array": Array [
    "",
    "",
  ],
  "field": "f1",
  "field2": "",
  "subform": Object {
    "field": "sub-f1",
    "field2": "",
  },
}
`;
